src/Symfony/Component/Messenger/Event/WorkerMessageFailedEvent.php

```diff
@@ -44,4 +45,14 @@ public function setForRetry(): void
     {
         $this->willRetry = true;
     }
+
+    public function addErrorDetailsStamp(ErrorDetailsStamp $stamp): void
+    {
+        // Do not append duplicate information
+        if ($stamp->equals($this->envelope->last(ErrorDetailsStamp::class))) {
+            return;
+        }
+
+        $this->envelope = $this->envelope->with($stamp);
+    }
```

Comment:
I'm not sure this needs to live in this class - we could just as easily put it inside of `AddErrorDetailsStampListener::onMessageFailed()`

Reply:
To do that, we would have to add a `setEnvelope` in order to override the existing envelope with the new stamp.

Reply:
Yes, I see it now. What about an `addStamp()` method on the event? I think that would be enough to avoid the `protected $envelope` and would allow us to move this method into `AddErrorDetailsStampListener`. You could then add stamps, but not entirely replace the envelope.

Reply:
👍  reverted back to the `addStamps()` method on the abstract parent.